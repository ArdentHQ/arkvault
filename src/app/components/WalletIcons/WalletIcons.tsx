import { constantCase } from "@payvo/sdk-helpers";
import { Contracts } from "@payvo/sdk-profiles";
import React from "react";
import { useTranslation } from "react-i18next";

import { Icon } from "@/app/components/Icon";
import { Tooltip } from "@/app/components/Tooltip";
import { Size } from "@/types";

interface WalletIconsProperties {
	exclude?: string[];
	iconColor?: string;
	iconSize?: Size;
	wallet: Contracts.IReadWriteWallet;
	tooltipDarkTheme?: boolean;
}

interface WalletIconProperties {
	type: "SecondSignature" | "Multisignature" | "TestNetwork" | "Ledger" | "Starred" | "Verified";
	label?: string;
	iconColor?: string;
	iconSize?: Size;
	tooltipDarkTheme?: boolean;
}

const getIconName = (type: string) => {
	if (type === "Starred") {
		return "StarFilled";
	}

	if (type === "Verified") {
		return "UserCheckMark";
	}

	if (type === "TestNetwork") {
		return "Code";
	}

	return type;
};

const getIconColor = (type: string) => (type === "Starred" ? "text-theme-warning-400" : "text-theme-secondary-700");

export const WalletIcon = ({ type, label, iconColor, iconSize = "lg", tooltipDarkTheme }: WalletIconProperties) => {
	const { t } = useTranslation();

	return (
		<Tooltip
			content={label || t(`COMMON.${constantCase(type)}` as const as any)}
			theme={tooltipDarkTheme ? "dark" : undefined}
		>
			<div data-testid={`WalletIcon__${type}`} className={`inline-block p-1 ${iconColor || getIconColor(type)}`}>
				<Icon name={getIconName(type)} size={iconSize} />
			</div>
		</Tooltip>
	);
};

export const WalletIcons = ({ exclude, wallet, ...iconProperties }: WalletIconsProperties) => {
	const { t } = useTranslation();

	return (
		<>
			{!exclude?.includes("isKnown") && wallet.isKnown() && (
				<WalletIcon
					type="Verified"
					label={t(`COMMON.VERIFIED`, { value: wallet.knownName() })}
					{...iconProperties}
				/>
			)}
			{!exclude?.includes("isSecondSignature") && wallet.hasSyncedWithNetwork() && wallet.isSecondSignature() && (
				<WalletIcon type="SecondSignature" label={t("COMMON.SECOND_SIGNATURE")} {...iconProperties} />
			)}
			{!exclude?.includes("isLedger") && wallet.isLedger() && <WalletIcon type="Ledger" {...iconProperties} />}
			{!exclude?.includes("isStarred") && wallet.isStarred() && <WalletIcon type="Starred" {...iconProperties} />}
			{!exclude?.includes("isMultiSignature") && wallet.hasSyncedWithNetwork() && wallet.isMultiSignature() && (
				<WalletIcon type="Multisignature" {...iconProperties} />
			)}
			{!exclude?.includes("isTestNetwork") && wallet.network().isTest() && (
				<WalletIcon type="TestNetwork" {...iconProperties} />
			)}
		</>
	);
};
