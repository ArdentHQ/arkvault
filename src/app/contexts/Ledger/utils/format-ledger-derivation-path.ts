import { LedgerDerivationScheme } from "@/app/contexts/Ledger/Ledger.contracts";

export const formatLedgerDerivationPath = (scheme: LedgerDerivationScheme) =>
	`m/${scheme.purpose || 44}'/${scheme.coinType}'/${scheme.account || 0}'/${scheme.change || 0}/${
		scheme.address || 0
	}`;
