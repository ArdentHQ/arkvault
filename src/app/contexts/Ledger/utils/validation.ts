import semver from "semver";
import { Coins } from "@payvo/sdk";
import { minVersionList } from "@/app/contexts/Ledger/Ledger.contracts";

export const hasRequiredAppVersion = async (coin: Coins.Coin) => {
	const coinId = coin.network().coin();

	if (minVersionList[coinId]) {
		const currentVersion = await coin.ledger().getVersion();

		if (semver.lt(currentVersion, minVersionList[coinId])) {
			return false;
		}
	}

	return true;
};
