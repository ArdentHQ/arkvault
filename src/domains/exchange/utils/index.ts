export * from "./assertions";
export * from "./errors";
