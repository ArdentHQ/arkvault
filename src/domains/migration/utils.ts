import { MigrationTransactionStatus } from "@/domains/migration/migration.contracts";

export const getIcon = (status: MigrationTransactionStatus) => {
	if (status === MigrationTransactionStatus.Confirmed) {
		return {
			color: "text-theme-success-600",
			name: "CircleCheckMark",
		};
	}

	return {
		color: "text-theme-warning-300",
		name: "Clock",
	};
};
