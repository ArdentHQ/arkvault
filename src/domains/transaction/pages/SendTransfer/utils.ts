export const getFeeType = (numberOfRecipients: number): "multiPayment" | "transfer" =>
	numberOfRecipients > 1 ? "multiPayment" : "transfer";
