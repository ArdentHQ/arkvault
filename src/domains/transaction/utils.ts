import { Services } from "@ardenthq/sdk";
import { RecipientItem } from "@/domains/transaction/components/RecipientList/RecipientList.contracts";

export const isNoDeviceError = (error: any) => {
	if (!error) {
		return false;
	}

	if (String(error).includes("no device found")) {
		return true;
	}

	return false;
};

export const isRejectionError = (error: any) => String(error).includes("Condition of use not satisfied");

export const handleBroadcastError = ({ errors }: Services.BroadcastResponse) => {
	const allErrors = Object.values(errors);

	if (allErrors.length === 0) {
		return;
	}

	throw new Error(allErrors[0]);
};

export const getTransferType = ({ recipients }: { recipients: RecipientItem[] }): "multiPayment" | "transfer" => {
	const isMultiPayment = recipients.length > 1;
	return isMultiPayment ? "multiPayment" : "transfer";
};

export const withAbortPromise =
	(signal?: AbortSignal, callback?: () => void) =>
	<T>(promise: Promise<T>) =>
		new Promise<T>((resolve, reject) => {
			if (signal) {
				signal.addEventListener("abort", () => {
					callback?.();
					reject("ERR_ABORT");
				});
			}

			return promise.then(resolve).catch(reject);
		});
