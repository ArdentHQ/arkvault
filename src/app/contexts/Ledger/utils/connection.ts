import { Coins } from "@ardenthq/sdk";
import retry, { AbortError, Options } from "p-retry";

import { formatLedgerDerivationPath } from "./format-ledger-derivation-path";
import { hasRequiredAppVersion } from "./validation";

export const accessLedgerDevice = async (coin: Coins.Coin) => {
	try {
		await coin.__construct();
		await coin.ledger().connect();
	} catch (error) {
		// If the device is open, continue normally.
		// Can be triggered when the user retries ledger connection.
		if (error.message !== "The device is already open.") {
			throw error;
		}
	}
};

export const accessLedgerApp = async ({ coin }: { coin: Coins.Coin }) => {
	await accessLedgerDevice(coin);

	if (!(await hasRequiredAppVersion(coin))) {
		throw new Error("VERSION_ERROR");
	}

	// Ensure that the app is accessible.
	await coin.ledger().getPublicKey(
		formatLedgerDerivationPath({
			coinType: coin.config().get<number>("network.constants.slip44"),
		}),
	);
};

export const persistLedgerConnection = async ({
	coin,
	options,
	hasRequestedAbort,
}: {
	coin: Coins.Coin;
	options: Options;
	hasRequestedAbort: () => boolean;
}) => {
	const retryAccess: any = async (attempts: number) => {
		if (hasRequestedAbort() && attempts > 1) {
			throw new AbortError("CONNECTION_ERROR");
		}

		try {
			await accessLedgerApp({ coin });
		} catch (error) {
			// Abort on version error or continue retrying access.
			if (error.message === "VERSION_ERROR") {
				throw new AbortError("VERSION_ERROR");
			}

			throw error;
		}
	};

	await retry(retryAccess, options);
};
