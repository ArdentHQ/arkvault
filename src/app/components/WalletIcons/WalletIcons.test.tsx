import { Contracts } from "@payvo/sdk-profiles";
import userEvent from "@testing-library/user-event";
import React from "react";

import { WalletIcons } from "./WalletIcons";
import { env, getDefaultProfileId, render, screen } from "@/utils/testing-library";

let wallet: Contracts.IReadWriteWallet;

describe("WalletIcons", () => {
	beforeEach(async () => {
		wallet = env.profiles().findById(getDefaultProfileId()).wallets().first();
		await wallet.synchroniser().identity();
	});

	it("should render with tooltip in the dark mode", () => {
		const walletSpy = jest.spyOn(wallet, "isKnown").mockReturnValue(true);

		render(<WalletIcons wallet={wallet} tooltipDarkTheme />);

		userEvent.hover(screen.getByTestId("WalletIcon__Verified"));

		expect(screen.getByRole("tooltip")).toHaveAttribute("data-theme", "dark");

		walletSpy.mockRestore();
	});

	it("should render the verified icon", () => {
		const walletSpy = jest.spyOn(wallet, "isKnown").mockReturnValue(true);

		render(<WalletIcons wallet={wallet} />);

		expect(screen.getByTestId("WalletIcon__Verified")).toBeInTheDocument();
		expect(screen.getByTestId("WalletIcon__Verified")).toHaveTextContent("user-check-mark.svg");

		walletSpy.mockRestore();
	});

	it("should render the ledger icon", () => {
		const walletSpy = jest.spyOn(wallet, "isLedger").mockReturnValue(true);

		render(<WalletIcons wallet={wallet} />);

		expect(screen.getByTestId("WalletIcon__Ledger")).toBeInTheDocument();
		expect(screen.getByTestId("WalletIcon__Ledger")).toHaveTextContent("ledger.svg");

		walletSpy.mockRestore();
	});

	it("should render the second signature icon", () => {
		const hasSyncedWithNetworkSpy = jest.spyOn(wallet, "hasSyncedWithNetwork").mockReturnValue(true);
		const walletSpy = jest.spyOn(wallet, "isSecondSignature").mockReturnValue(true);

		render(<WalletIcons wallet={wallet} />);

		expect(screen.getByTestId("WalletIcon__SecondSignature")).toBeInTheDocument();

		walletSpy.mockRestore();
		hasSyncedWithNetworkSpy.mockRestore();
	});

	it("should render the star icon", () => {
		const walletSpy = jest.spyOn(wallet, "isStarred").mockReturnValue(true);

		render(<WalletIcons wallet={wallet} />);

		expect(screen.getByTestId("WalletIcon__Starred")).toBeInTheDocument();
		expect(screen.getByTestId("WalletIcon__Starred")).toHaveTextContent("star-filled.svg");

		walletSpy.mockRestore();
	});

	it("should render the multisignature icon", () => {
		const hasSyncedWithNetworkSpy = jest.spyOn(wallet, "hasSyncedWithNetwork").mockReturnValue(true);
		const isMultiSignatureSpy = jest.spyOn(wallet, "isMultiSignature").mockReturnValue(true);

		render(<WalletIcons wallet={wallet} />);

		expect(screen.getByTestId("WalletIcon__Multisignature")).toBeInTheDocument();
		expect(screen.getByTestId("WalletIcon__Multisignature")).toHaveTextContent("multisignature.svg");

		hasSyncedWithNetworkSpy.mockRestore();
		isMultiSignatureSpy.mockRestore();
	});

	it("should render the test network icon", () => {
		const walletSpy = jest.spyOn(wallet.network(), "isTest").mockReturnValue(true);

		render(<WalletIcons wallet={wallet} />);

		expect(screen.getByTestId("WalletIcon__TestNetwork")).toBeInTheDocument();
		expect(screen.getByTestId("WalletIcon__TestNetwork")).toHaveTextContent("code.svg");

		walletSpy.mockRestore();
	});

	it("should not render excluded icons", () => {
		const walletSpy = jest.spyOn(wallet, "isStarred").mockReturnValue(true);

		const { container } = render(<WalletIcons wallet={wallet} exclude={["isStarred"]} />);

		expect(screen.queryByTestId("WalletIcon__Starred")).not.toBeInTheDocument();
		expect(container).not.toHaveTextContent("star-filled.svg");

		walletSpy.mockRestore();
	});
});
