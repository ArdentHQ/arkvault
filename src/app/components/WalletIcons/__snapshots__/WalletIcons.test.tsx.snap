// Vitest Snapshot v1

exports[`WalletIcons > should render wallet icon skeleton 1`] = `
<DocumentFragment>
  <span
    aria-busy="true"
    aria-live="polite"
    class="flex w-auto max-w-full items-center overflow-hidden leading-none"
  >
    <span
      class="react-loading-skeleton"
      style="width: 54px; height: 20px;"
    >
      â€Œ
    </span>
    <br />
  </span>
</DocumentFragment>
`;
